README.md
<div>
    <a href="https://travis-ci.org/pjfl/p5-dist-zilla-plugin-manifestinroot"><img src="https://travis-ci.org/pjfl/p5-dist-zilla-plugin-manifestinroot.svg?branch=master" alt="Travis CI Badge"></a>
    <a href="http://badge.fury.io/pl/Dist-Zilla-Plugin-ManifestInRoot"><img src="https://badge.fury.io/pl/Dist-Zilla-Plugin-ManifestInRoot.svg" alt="CPAN Badge"></a>
    <a href="http://cpants.cpanauthors.org/dist/Dist-Zilla-Plugin-ManifestInRoot"><img src="http://cpants.cpanauthors.org/dist/Dist-Zilla-Plugin-ManifestInRoot.png" alt="Kwalitee Badge"></a>
</div>

# Name

Dist::Zilla::Plugin::ManifestInRoot - Puts the MANIFEST file in the project root

# Synopsis

    # In your dist.ini
    [ManifestInRoot]

# Version

This documents version v0.16.$Rev: 1 $ of [Dist::Zilla::Plugin::ManifestInRoot](https://metacpan.org/pod/Dist::Zilla::Plugin::ManifestInRoot)

# Description

Puts the `MANIFEST` file in the project root so that it can be used by
other programs, e.g. [App::Cpanminus](https://metacpan.org/pod/App::Cpanminus) and [Module::Provision](https://metacpan.org/pod/Module::Provision)

# Configuration and Environment

Defined these attributes;

- `excludes`

    An array reference of file names to exclude from the root directory copy of the
    manifest.  Set using the multivalued initialisation argument `exclude`.
    Defaults to the list of files generated by the build process

# Subroutines/Methods

## after\_build

Copy the `MANIFEST` file from the build directory to the project root filtering
out unwanted files

## before\_build

Delete the existing `MANIFEST` file

## gather\_files

Create the content providing callback

## `mvp_multivalue_args`

Returns a list of configuration attribute names that are treated as
multi valued

# Diagnostics

None

# Dependencies

- [Dist::Zilla::Role::InstallTool](https://metacpan.org/pod/Dist::Zilla::Role::InstallTool)
- [File::Slurp](https://metacpan.org/pod/File::Slurp)
- [Moose](https://metacpan.org/pod/Moose)

# Incompatibilities

There are no known incompatibilities in this module

# Bugs and Limitations

There are no known bugs in this module. Please report problems to
http://rt.cpan.org/NoAuth/Bugs.html?Dist=Dist-Zilla-Plugin-ManifestInRoot.
Patches are welcome

# Acknowledgements

Larry Wall - For the Perl programming language

# Author

Peter Flanigan, `<pjfl@cpan.org>`

# License and Copyright

Copyright (c) 2015 Peter Flanigan. All rights reserved

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself. See [perlartistic](https://metacpan.org/pod/perlartistic)

This program is distributed in the hope that it will be useful,
but WITHOUT WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE


Changes
Revision history for Dist::Zilla::Plugin::ManifestInRoot

          - Use Path::Tiny->remove to unlink existing file
            This is the way it should have been done all along
            works for both D:Z using Path::Class and D:Z using
            Path::Tiny

0.16.1    2016-12-29 18:46:31
          - Avoid Path::Tiny and Path::Class, use File::Spec
            Total cockup in backward compatibility

0.15.1    2016-12-25 16:13:44
          - Tests only for developers

0.14.1    2016-12-24 21:10:37
          - Toolchain update
          - Refactored smoker exceptions to dump file
          - Fewer test dependencies in boilerplate

0.13.1    2015-09-26 01:04:22
          - Old smokers may not understand TestRequires

0.12.1    2015-09-15 02:41:08
          - Broken smoker 85ab240c-6bfd-1014-828b-aa2bd4cf89d2
          - Toolchain update
          - Better prereq data for build and test phases

0.11.1    2015-08-29 10:01:39
          - Broken smoker 85ab240c-6bfd-1014-828b-aa2bd4cf89d2

0.10.1    2015-08-03 13:28:29
          - Fixed tests on mwwin32

0.9.1     2015-08-02 18:14:02
          - Fixed line termination on mswin32

0.8.1     2015-08-02 12:07:20
          - Can filter unwanted files from root copy
          - Updated toolchain

0.7.1     2015-05-22 09:24:18
          - Added badges and public repository

0.6.1     2014-08-27 16:52:09
          - Hopefully this time the broken smoker will be silenced
            b6c8672d-6bf5-1014-99ad-e1a5890cd912

0.5.1     2014-08-26 18:02:12
          - This broken smoker is a rash across CPAN
            69bc01aa-6c01-1014-a840-c829b95c3ac1

0.4.1     2014-05-17 12:47:04
          - Copy MANIFEST rather than read / write

0.3.1     2014-05-15 22:25:04
          - Removed dependency on File::Slurp

0.2.1     2013-08-15 11:17:02
          - Switched to Dist::Zilla

0.1.3     2013-08-07 14:47:32
          - Reset min perl testing to original behaviour
          - Added code from Module::Provision
          - Added project files to master branch



t/boilerplate.pm
package t::boilerplate;

use strict;
use warnings;
use File::Spec::Functions qw( catdir updir );
use FindBin               qw( $Bin );
use lib               catdir( $Bin, updir, 'lib' ), catdir( $Bin, 'lib' );

use Module::Build;
use Sys::Hostname;

sub plan (;@) {
   $_[ 0 ] eq 'skip_all' and print '1..0 # SKIP '.$_[ 1 ]."\n" and exit 0;
}

my ($builder, $host, $notes, $perl_ver);

BEGIN {
   $host     = lc hostname;
   $builder  = eval { Module::Build->current };
   $notes    = $builder ? $builder->notes : {};
   $perl_ver = $notes->{min_perl_version} || 5.008;

   eval { require Test::Requires }; $@ and plan skip_all => 'No Test::Requires';

   $Bin =~ m{ : .+ : }mx and plan skip_all => 'Two colons in $Bin path';
   # Not possible to detect smoking
   # Smoker has two colons in Dist::Zilla::Test rootdir path
   $host eq 'w5050029' and plan skip_all => 'Broken smoker';
   $host eq 'w5139020' and plan skip_all =>
      'Broken smoker 85ab240c-6bfd-1014-828b-aa2bd4cf89d2';

   if ($notes->{testing}) {
   }
}

use Test::Requires "${perl_ver}";
use Test::Requires { version => 0.88 };

sub import {
   strict->import;
   $] < 5.008 ? warnings->import : warnings->import( NONFATAL => 'all' );
   return;
}

1;

# Local Variables:
# mode: perl
# tab-width: 3
# End:
# vim: expandtab shiftwidth=3:


.build.PL
use strict;
use warnings;

use Module::Build ##{ $meta2{prereqs}->{configure}->{requires}->{'Module::Build'} ##};

my $perl_ver = ##{ $meta2{prereqs}->{runtime}->{requires}->{perl} ##};
my $name     = '##{ $zilla->name ##}'; $name =~ s{ - }{::}gmx;
my $prereqs  = {
##{ $plugin->get_prereqs( 1 ) ##}
};
my $testing  = sub {   $ENV{AUTOMATED_TESTING}
                  ||   $ENV{PERL_CPAN_REPORTER_CONFIG}
                  ||   $ENV{PERL5_CPANPLUS_IS_VERSION}
                  || (($ENV{TEMP} || q()) =~ m{ a00772a }mx) };
my $whimper  = sub { print {*STDOUT} $_[ 0 ]."\n"; exit 0 };
my $stop     = sub {
   $testing->() or $whimper->( $_[ 0 ] );
   $prereqs->{build_requires} = { %{ $prereqs->{configure_requires} } };
   $prereqs->{test_requires } = { %{ $prereqs->{configure_requires} } };
   $prereqs->{test_requires }->{ 'Test::Requires' } = 0.06;
   delete $prereqs->{requires};
};

# Not interested in reports below min perl ver
$] < $perl_ver and $stop->( "Minimum required Perl version is ${perl_ver}" );

Module::Build->new(
   'module_name' => $name,
   'notes'       => { min_perl_version => $perl_ver, testing => $testing->() },
   %{ $prereqs },
##{ $plugin->get_default('share_dir') ##}
)->create_build_script;

exit 0;


Build.PL
use strict;
use warnings;

use Module::Build 0.4004;

my $perl_ver = 5.008;
my $name     = 'Dist-Zilla-Plugin-ManifestInRoot'; $name =~ s{ - }{::}gmx;
my $prereqs  = {
  'build_requires' => {
    'Module::Build' => '0.4004',
    'version' => '0.88'
  },
  'configure_requires' => {
    'Module::Build' => '0.4004',
    'version' => '0.88'
  },
  'requires' => {
    'Dist::Zilla' => '4.300033',
    'Moose' => '2.0802',
    'Path::Tiny' => '0.068',
    'namespace::autoclean' => '0.22',
    'perl' => '5.008',
    'version' => '0.88'
  },
  'test_requires' => {
    'Dist::Zilla' => '4.300033',
    'File::Spec' => '0',
    'Module::Build' => '0.4004',
    'Module::Metadata' => '0',
    'Sys::Hostname' => '0',
    'Test::Requires' => '0.06',
    'version' => '0.88'
  },
};
my $testing  = sub {   $ENV{AUTOMATED_TESTING}
                  ||   $ENV{PERL_CPAN_REPORTER_CONFIG}
                  ||   $ENV{PERL5_CPANPLUS_IS_VERSION}
                  || (($ENV{TEMP} || q()) =~ m{ a00772a }mx) };
my $whimper  = sub { print {*STDOUT} $_[ 0 ]."\n"; exit 0 };
my $stop     = sub {
   $testing->() or $whimper->( $_[ 0 ] );
   $prereqs->{build_requires} = { %{ $prereqs->{configure_requires} } };
   $prereqs->{test_requires } = { %{ $prereqs->{configure_requires} } };
   $prereqs->{test_requires }->{ 'Test::Requires' } = 0.06;
   delete $prereqs->{requires};
};

# Not interested in reports below min perl ver
$] < $perl_ver and $stop->( "Minimum required Perl version is ${perl_ver}" );

Module::Build->new(
   'module_name' => $name,
   'notes'       => { min_perl_version => $perl_ver, testing => $testing->() },
   %{ $prereqs },

)->create_build_script;

exit 0;
